File: frontend/src/shared/lib/indicators.ts
/** 日线重采样与常见技术指标（MA / MACD / KDJ / RSI / BOLL 等）。 */

export interface OhlcBar {
  trade_date: string
  open: number | null
  high: number | null
  low: number | null
  close: number | null
  volume: number | null
  amount?: number | null
  turnover?: number | null
}

export type KPeriod = 'day' | 'week' | 'month'

function num(value: number | null | undefined): number | null {
  return value === null || value === undefined || Number.isNaN(Number(value))
    ? null
    : Number(value)
}

function weekKey(date: string): string {
  const d = new Date(`${date}T00:00:00`)
  const day = (d.getDay() + 6) % 7 // Mon=0
  d.setDate(d.getDate() - day)
  return d.toISOString().slice(0, 10)
}

function monthKey(date: string): string {
  return date.slice(0, 7)
}

/** 由日线合成周/月 K：开=首、高=max、低=min、收=末，量额换手率求和。 */
export function resampleBars(bars: OhlcBar[], period: KPeriod): OhlcBar[] {
  if (period === 'day' || bars.length === 0) return bars
  const keyOf = period === 'week' ? weekKey : monthKey
  const groups = new Map<string, OhlcBar[]>()
  for (const bar of bars) {
    const key = keyOf(bar.trade_date)
    const bucket = groups.get(key)
    if (bucket) bucket.push(bar)
    else groups.set(key, [bar])
  }
  const out: OhlcBar[] = []
  for (const bucket of groups.values()) {
    const first = bucket[0]
    const last = bucket[bucket.length - 1]
    let high = -Infinity
    let low = Infinity
    let volume = 0
    let amount = 0
    let hasVol = false
    let hasAmt = false
    let turnover = 0
    let hasTurnover = false
    for (const bar of bucket) {
      const h = num(bar.high)
      const l = num(bar.low)
      if (h !== null) high = Math.max(high, h)
      if (l !== null) low = Math.min(low, l)
      const v = num(bar.volume)
      if (v !== null) {
        volume += v
        hasVol = true
      }
      const a = num(bar.amount)
      if (a !== null) {
        amount += a
        hasAmt = true
      }
      const t = num(bar.turnover)
      if (t !== null) {
        turnover += t
        hasTurnover = true
      }
    }
    out.push({
      trade_date: last.trade_date,
      open: num(first.open),
      high: high === -Infinity ? null : high,
      low: low === Infinity ? null : low,
      close: num(last.close),
      volume: hasVol ? volume : null,
      amount: hasAmt ? amount : null,
      turnover: hasTurnover ? turnover : null,
    })
  }
  return out
}

export function sma(values: Array<number | null>, period: number): Array<number | null> {
  const out: Array<number | null> = Array(values.length).fill(null)
  let sum = 0
  let count = 0
  const queue: Array<number | null> = []
  for (let i = 0; i < values.length; i += 1) {
    const v = values[i]
    queue.push(v)
    if (v !== null) {
      sum += v
      count += 1
    }
    if (queue.length > period) {
      const old = queue.shift()
      if (old !== null && old !== undefined) {
        sum -= old
        count -= 1
      }
    }
    out[i] = queue.length === period && count === period ? sum / period : null
  }
  return out
}

export function ema(values: Array<number | null>, period: number): Array<number | null> {
  const out: Array<number | null> = Array(values.length).fill(null)
  const k = 2 / (period + 1)
  let prev: number | null = null
  for (let i = 0; i < values.length; i += 1) {
    const v = values[i]
    if (v === null) {
      out[i] = null
      continue
    }
    prev = prev === null ? v : v * k + prev * (1 - k)
    out[i] = prev
  }
  return out
}

export interface MacdSeries {
  dif: Array<number | null>
  dea: Array<number | null>
  hist: Array<number | null>
}

export function macd(
  closes: Array<number | null>,
  fast = 12,
  slow = 26,
  signal = 9,
): MacdSeries {
  const emaFast = ema(closes, fast)
  const emaSlow = ema(closes, slow)
  const dif = closes.map((_, i) => {
    const a = emaFast[i]
    const b = emaSlow[i]
    return a === null || b === null ? null : a - b
  })
  const dea = ema(dif, signal)
  const hist = dif.map((d, i) => {
    const e = dea[i]
    return d === null || e === null ? null : (d - e) * 2
  })
  return { dif, dea, hist }
}

export interface KdjSeries {
  k: Array<number | null>
  d: Array<number | null>
  j: Array<number | null>
}

export function kdj(
  highs: Array<number | null>,
  lows: Array<number | null>,
  closes: Array<number | null>,
  n = 9,
  m1 = 3,
  m2 = 3,
): KdjSeries {
  const rsv: Array<number | null> = Array(closes.length).fill(null)
  for (let i = 0; i < closes.length; i += 1) {
    if (i + 1 < n) continue
    const start = i - n + 1
    let hh = -Infinity
    let ll = Infinity
    let ok = true
    for (let j = start; j <= i; j += 1) {
      const h = highs[j]
      const l = lows[j]
      if (h === null || l === null) {
        ok = false
        break
      }
      hh = Math.max(hh, h)
      ll = Math.min(ll, l)
    }
    const c = closes[i]
    if (!ok || c === null || hh === ll) {
      rsv[i] = ok && c !== null && hh === ll ? 50 : null
      continue
    }
    rsv[i] = ((c - ll) / (hh - ll)) * 100
  }

  const k: Array<number | null> = Array(closes.length).fill(null)
  const d: Array<number | null> = Array(closes.length).fill(null)
  const j: Array<number | null> = Array(closes.length).fill(null)
  let prevK: number | null = null
  let prevD: number | null = null
  for (let i = 0; i < closes.length; i += 1) {
    const r = rsv[i]
    if (r === null) continue
    const curK: number = prevK === null ? r : (prevK * (m1 - 1) + r) / m1
    const curD: number = prevD === null ? curK : (prevD * (m2 - 1) + curK) / m2
    k[i] = curK
    d[i] = curD
    j[i] = 3 * curK - 2 * curD
    prevK = curK
    prevD = curD
  }
  return { k, d, j }
}

function rollingMeanDeviation(values: Array<number | null>, period: number): Array<number | null> {
  const out: Array<number | null> = Array(values.length).fill(null)
  for (let i = period - 1; i < values.length; i += 1) {
    const window = values.slice(i - period + 1, i + 1)
    if (window.some((value) => value === null)) continue
    const mean = window.reduce<number>((sum, value) => sum + (value as number), 0) / period
    out[i] = window.reduce<number>((sum, value) => sum + Math.abs((value as number) - mean), 0) / period
  }
  return out
}

function rollingStd(values: Array<number | null>, period: number): Array<number | null> {
  const out: Array<number | null> = Array(values.length).fill(null)
  for (let i = period - 1; i < values.length; i += 1) {
    const window = values.slice(i - period + 1, i + 1)
    if (window.some((value) => value === null)) continue
    const mean = window.reduce<number>((sum, value) => sum + (value as number), 0) / period
    const squared = window.reduce<number>((sum, value) => sum + ((value as number) - mean) ** 2, 0)
    out[i] = period > 1 ? Math.sqrt(squared / (period - 1)) : null
  }
  return out
}

export function rsi(values: Array<number | null>, period = 14): Array<number | null> {
  const out: Array<number | null> = Array(values.length).fill(null)
  let previous: number | null = null
  let averageGain: number | null = null
  let averageLoss: number | null = null
  let validCloses = 0
  for (let i = 0; i < values.length; i += 1) {
    const current = values[i]
    if (current === null) continue
    validCloses += 1
    if (previous === null) {
      previous = current
      continue
    }
    const delta = current - previous
    const gain = Math.max(delta, 0)
    const loss = Math.max(-delta, 0)
    averageGain = averageGain === null ? gain : ((period - 1) * averageGain + gain) / period
    averageLoss = averageLoss === null ? loss : ((period - 1) * averageLoss + loss) / period
    previous = current
    if (validCloses < period + 1) continue
    const total = averageGain + averageLoss
    out[i] = total === 0 ? 50 : (averageGain / total) * 100
  }
  return out
}

export function trueRange(
  highs: Array<number | null>,
  lows: Array<number | null>,
  closes: Array<number | null>,
): Array<number | null> {
  const out: Array<number | null> = Array(closes.length).fill(null)
  for (let i = 1; i < closes.length; i += 1) {
    const high = highs[i]
    const low = lows[i]
    const previousClose = closes[i - 1]
    if (high === null || low === null || previousClose === null) continue
    out[i] = Math.max(high - low, Math.abs(high - previousClose), Math.abs(low - previousClose))
  }
  return out
}

export function atr(
  highs: Array<number | null>,
  lows: Array<number | null>,
  closes: Array<number | null>,
  period: number,
): Array<number | null> {
  return sma(trueRange(highs, lows, closes), period)
}

export function roc(values: Array<number | null>, period: number): Array<number | null> {
  return values.map((value, index) => {
    const previous = index >= period ? values[index - period] : null
    if (value === null || previous === null || previous === 0) return null
    return ((value - previous) / previous) * 100
  })
}

export function williamsR(
  highs: Array<number | null>,
  lows: Array<number | null>,
  closes: Array<number | null>,
  period: number,
): Array<number | null> {
  const out: Array<number | null> = Array(closes.length).fill(null)
  for (let i = period - 1; i < closes.length; i += 1) {
    const highWindow = highs.slice(i - period + 1, i + 1)
    const lowWindow = lows.slice(i - period + 1, i + 1)
    const close = closes[i]
    if (close === null || highWindow.some((value) => value === null) || lowWindow.some((value) => value === null)) continue
    const highest = Math.max(...(highWindow as number[]))
    const lowest = Math.min(...(lowWindow as number[]))
    out[i] = highest === lowest ? 0 : -((highest - close) / (highest - lowest)) * 100
  }
  return out
}

export function cci(
  highs: Array<number | null>,
  lows: Array<number | null>,
  closes: Array<number | null>,
  period: number,
): Array<number | null> {
  const typical = closes.map((close, index) => {
    const high = highs[index]
    const low = lows[index]
    return high === null || low === null || close === null ? null : (high + low + close) / 3
  })
  const mean = sma(typical, period)
  const deviation = rollingMeanDeviation(typical, period)
  return typical.map((value, index) => {
    const base = mean[index]
    const dev = deviation[index]
    if (value === null || base === null || dev === null) return null
    return dev === 0 ? 0 : (value - base) / (0.015 * dev)
  })
}

export function obv(
  closes: Array<number | null>,
  volumes: Array<number | null>,
): Array<number | null> {
  const out: Array<number | null> = Array(closes.length).fill(null)
  let total = 0
  for (let i = 0; i < closes.length; i += 1) {
    const close = closes[i]
    const volume = volumes[i]
    if (close === null || volume === null) continue
    const previous = i > 0 ? closes[i - 1] : null
    if (previous !== null) {
      if (close > previous) total += volume
      else if (close < previous) total -= volume
    }
    out[i] = total
  }
  return out
}

export interface BollingerSeries {
  mid: Array<number | null>
  upper: Array<number | null>
  lower: Array<number | null>
}

export function bollinger(
  values: Array<number | null>,
  period = 20,
  deviations = 2,
): BollingerSeries {
  const mid = sma(values, period)
  const std = rollingStd(values, period)
  return {
    mid,
    upper: std.map((value, index) => (value === null || mid[index] === null ? null : mid[index] + value * deviations)),
    lower: std.map((value, index) => (value === null || mid[index] === null ? null : mid[index] - value * deviations)),
  }
}

File: src/formula/domain/indicators.py
"""常用技术指标的向量化实现。

这些函数只负责数值计算，不负责行情加载或信号裁决。Series 与 DataFrame
共用同一条路径，保证单票调试结果可以直接推广到全市场面板。
"""
from __future__ import annotations

from typing import TypeAlias

import numpy as np
import pandas as pd

from src.formula.domain.functions import (
    ABS,
    AVEDEV,
    EMA,
    HHV,
    LLV,
    MA,
    MAX,
    REF,
    SMA,
    STD,
)

Frame: TypeAlias = pd.Series | pd.DataFrame

__all__ = [
    "ATR",
    "BOLL_LOWER",
    "BOLL_MID",
    "BOLL_UPPER",
    "CCI",
    "MACD",
    "MACD_DEA",
    "MACD_DIF",
    "OBV",
    "ROC",
    "RSI",
    "TR",
    "WR",
]


def _require_period(periods: int, name: str) -> int:
    if isinstance(periods, bool) or not isinstance(periods, int) or periods <= 0:
        raise ValueError(f"{name} 的周期必须是正整数")
    return periods


def _valid_prefix(source: Frame, minimum: int) -> Frame:
    valid = source.notna().cumsum(axis=0)
    return valid >= minimum


def _mask_prefix(result: Frame, source: Frame, minimum: int) -> Frame:
    return result.where(_valid_prefix(source, minimum))


def _safe_divide(numerator: Frame, denominator: Frame, fallback: float | None = None) -> Frame:
    with np.errstate(divide="ignore", invalid="ignore"):
        result = numerator / denominator
    if fallback is not None:
        result = result.where(denominator != 0, fallback)
    return result


def TR(high: Frame, low: Frame, close: Frame) -> Frame:
    """TR(HIGH, LOW, CLOSE)：真实波幅。首根因缺少昨收而为空。"""
    previous_close = REF(close, 1)
    return MAX(high - low, MAX(ABS(high - previous_close), ABS(low - previous_close)))


def ATR(high: Frame, low: Frame, close: Frame, periods: int) -> Frame:
    """ATR(HIGH, LOW, CLOSE, N)：N 周期真实波幅均值。"""
    periods = _require_period(periods, "ATR")
    return MA(TR(high, low, close), periods)


def RSI(series: Frame, periods: int = 14) -> Frame:
    """RSI(X, N)：按通达信 SMA 递推口径计算的相对强弱指标（0-100）。"""
    periods = _require_period(periods, "RSI")
    delta = series - REF(series, 1)
    gain = delta.clip(lower=0)
    loss = (-delta).clip(lower=0)
    average_gain = SMA(gain, periods, 1.0)
    average_loss = SMA(loss, periods, 1.0)
    total = average_gain + average_loss
    result = _safe_divide(average_gain * 100.0, total, fallback=50.0)
    return _mask_prefix(result, series, periods + 1)


def ROC(series: Frame, periods: int) -> Frame:
    """ROC(X, N)：N 周期价格变化率，单位为百分比。"""
    periods = _require_period(periods, "ROC")
    previous = REF(series, periods)
    return _safe_divide(series - previous, previous) * 100.0


def WR(high: Frame, low: Frame, close: Frame, periods: int) -> Frame:
    """WR(HIGH, LOW, CLOSE, N)：威廉指标，返回 -100 到 0。"""
    periods = _require_period(periods, "WR")
    highest = HHV(high, periods)
    lowest = LLV(low, periods)
    return _safe_divide((highest - close) * -100.0, highest - lowest, fallback=0.0)


def CCI(high: Frame, low: Frame, close: Frame, periods: int) -> Frame:
    """CCI(HIGH, LOW, CLOSE, N)：典型价格商品通道指标。"""
    periods = _require_period(periods, "CCI")
    typical = (high + low + close) / 3.0
    mean = MA(typical, periods)
    deviation = AVEDEV(typical, periods)
    return _safe_divide(typical - mean, deviation * 0.015, fallback=0.0)


def OBV(close: Frame, volume: Frame) -> Frame:
    """OBV(CLOSE, VOL)：按收盘涨跌方向累计成交量。"""
    change = close - REF(close, 1)
    direction = change.gt(0).astype(float) - change.lt(0).astype(float)
    valid = close.notna() & volume.notna()
    result = (volume.where(valid, 0.0) * direction.where(valid, 0.0)).cumsum(axis=0)
    return result.where(valid)


def MACD_DIF(close: Frame, fast: int = 12, slow: int = 26) -> Frame:
    """MACD_DIF(CLOSE, FAST, SLOW)：快慢 EMA 差值。"""
    fast = _require_period(fast, "MACD fast")
    slow = _require_period(slow, "MACD slow")
    if fast >= slow:
        raise ValueError("MACD fast 必须小于 slow")
    return EMA(close, fast) - EMA(close, slow)


def MACD_DEA(close: Frame, fast: int = 12, slow: int = 26, signal: int = 9) -> Frame:
    """MACD_DEA(CLOSE, FAST, SLOW, SIGNAL)：DIF 的信号线。"""
    signal = _require_period(signal, "MACD signal")
    return EMA(MACD_DIF(close, fast, slow), signal)


def MACD(close: Frame, fast: int = 12, slow: int = 26, signal: int = 9) -> Frame:
    """MACD(CLOSE, FAST, SLOW, SIGNAL)：(DIF-DEA)*2 柱值。"""
    dif = MACD_DIF(close, fast, slow)
    dea = EMA(dif, _require_period(signal, "MACD signal"))
    return (dif - dea) * 2.0


def BOLL_MID(series: Frame, periods: int) -> Frame:
    """BOLL_MID(X, N)：布林中轨。"""
    return MA(series, _require_period(periods, "BOLL"))


def BOLL_UPPER(series: Frame, periods: int, deviations: float = 2.0) -> Frame:
    """BOLL_UPPER(X, N, K)：布林上轨。"""
    periods = _require_period(periods, "BOLL")
    return MA(series, periods) + STD(series, periods) * float(deviations)


def BOLL_LOWER(series: Frame, periods: int, deviations: float = 2.0) -> Frame:
    """BOLL_LOWER(X, N, K)：布林下轨。"""
    periods = _require_period(periods, "BOLL")
    return MA(series, periods) - STD(series, periods) * float(deviations)
